_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,386 @@
+  schemas
+
+Stitching is one of the end-goals, but merging may be enough for a lot of cases. Both of the two launch posts above
+give a much more in-depth explanation of how everything comes together, but these should be enough for this post.
+
+## How Do We Do It?
+
+We have 5 GraphQL APIs inside the Artsy ecosystem, our aim is to cautiously include these APIs inside metaphysics.
+We don't need the entire contents of those APIs, and as you'll learn - we couldn't do that even if we wanted.
+
+The technique we settled on was:
+
+1.  Download the schema of each external API into metaphysics' source code
+1.  Have each schema trimmed to just the essentials that we need today
+1.  Merge in each schema incrementally
+1.  Stitch in any desired schema changes
+
+Let's dig, with some code into how we do each of these steps.
+
+#### Downloading Schemas
+
+We created a [pretty minimal script][dl-schema] which can be run periodically from a developer's computer.
+
+```js
+const destination = "src/data"
+
+const httpConvectionLink = createHttpLink({
+  fetch,
+  uri: urljoin("https://convection-staging.artsy.net/api", "graphql")
+})
+
+introspectSchema(httpConvectionLink).then(schema => {
+  fs.writeFileSync(
+    path.join(destination, "convection.graphql"),
+    printSchema(schema, { commentDescriptions: true })
+  )
+})
+```
+
+The script uses an [apollo-http-link][] to grab our schema, and store it in our repo, see
+[`src/data/convection.graphql`][c-gql]. This means that when someone wants to update to a new version of the
+schema, it will go through code review and a normal testing-flow. The trade-off being that it will always be out of
+date a little bit, but you can make guarantees about the current schema. This is a reasonable trade-off, as GraphQL
+schemas should always be forward compatible for queries, and when someone wants to use a new field from another
+service they can move the schema definition from [the git repo][rfc31].
+
+This file is the [GraphQL SDL][sdl] representations of the entire type system for that schema. This means we have a
+local copy of the schemas, so we can use it for tests for the next few steps.
+
+#### Schema Manipulation
+
+Each API writes for their own domain. This can be problematic when you use a `User` in one API, which isn't generic
+enough to be a `User` in a global API of all services combined. When thinking about this problem, we created a
+[guide for ourselves][guides-schema] on how to think about schema design at local and global level.
+
+We use a few of [the transform APIs][transform] available in graphql-tools to make the merges work. The first
+approach is to force a namespace by prefixing the merged Types with their domain.
+
+```js
+export const executableConvectionSchema = async () => {
+  const convectionLink = createConvectionLink()
+  const convectionTypeDefs = readFileSync("src/data/convection.graphql", "utf8")
```

Comment:
Is it weird that this is `readFileSync` inside an `async` function?

Reply:
`readFile` doesn't return a promise, so can't be used async, so nah

Reply:
Ha, gotcha.